### ISSUE TITLE
Immersive Mode not working after onWindowFocusChanged()

### ISSUE THREAD

**Issue by ConteZero** [libgdx/libgdx#3335]
In AndroidApplication.java:

``` java
    protected void hideStatusBar (boolean hide) {
        if (!hide || getVersion() < 11) return;

        View rootView = getWindow().getDecorView();

        try {
            Method m = View.class.getMethod("setSystemUiVisibility", int.class);
            if (getVersion() <= 13) m.invoke(rootView, 0x0);
            m.invoke(rootView, 0x1);
        } catch (Exception e) {
            log("AndroidApplication", "Can't hide status bar", e);
        }
    }

    @Override
    public void onWindowFocusChanged (boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        useImmersiveMode(this.useImmersiveMode);
        hideStatusBar(this.hideStatusBar);
        if (hasFocus) {
            this.wasFocusChanged = 1;
            if (this.isWaitingForAudio) {
                this.audio.resume();
                this.isWaitingForAudio = false;
            }
        } else {
            this.wasFocusChanged = 0;
        }
    }

    @TargetApi(19)
    @Override
    public void useImmersiveMode (boolean use) {
        if (!use || getVersion() < Build.VERSION_CODES.KITKAT) return;

        View view = getWindow().getDecorView();
        try {
            Method m = View.class.getMethod("setSystemUiVisibility", int.class);
            int code = View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION | View.SYSTEM_UI_FLAG_FULLSCREEN
                | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY;
            m.invoke(view, code);
        } catch (Exception e) {
            log("AndroidApplication", "Can't set immersive mode", e);
        }
    }
```

Method onWindowFocusChanged() contains a bug, useImmersiveMode() should be executed after hideStatusBar() and not before.
As it is now, hideStatusBar() overwrites the changes to SystemUiVisibility done by useImmersiveMode(), and this makes Immersive Mode not working after a call to onWindowFocusChanged()

**Comment by obigu** [comment]
What you point out here makes sense in paper, the call to setImmersiveMode seems to have no effect in onWindowFocusChanged.

I've ran tests on a game by switching both with a official and modified libGDX 1.7.0 with
```
seImmersiveMode(this.useImmersiveMode);
hideStatusBar(this.hideStatusBar);
```
and the modified
```
hideStatusBar(this.hideStatusBar);
seImmersiveMode(this.useImmersiveMode);
```

The test game connects to Google Play Services, has Facebook integration, external links and I've also played with volume keys, sleep key, etc... for a while.

During the tests I haven't managed to get a buggy sticky immersive mode in the game with the current code. On the other hand I've managed to bug it (get sticky buttons bar inside game) using the modified one that switches the order.

I'm not able to draw a conclusion as I don't know enough about the inners of activity lifecycles and order of listener calls but it might be that useImmersiveMode should Not be called in onWindowFocusChanged and it simply works because of the hideStatusBar after it that overrides its effect.

Do you have a scenario/test to be able to reproduce the bug?

**Comment by ConteZero** [comment]
I've tested with a Motorola Moto E 2015 (android 5.0) and if you press home button and than you return to the game the bar remain visible and do not disappear.
If I switch the order of these lines all works as expected.
Google says that you must reset Immersive Mode in onWindowFocusChanged:
[https://developer.android.com/training/system-ui/immersive.html](https://developer.android.com/training/system-ui/immersive.html)
I think that actual code is wrong, but it is probable that some devices have a faulty implementation of immersive mode and some more workarounds will be needed.

**Comment by obigu** [comment]
I agree, it looks like Immersive mode should also be set in onWindowFocusChanged according to documentation. My tests on a Nexus 4 (android 5.1) though show the opposite behaviour. I've redone them to verify and I've been able to bug it and cause the bars to remain visible bu using the modified version (using the sleep key when trying to log into Facebook and then returning to the game). As you say there may be inconsistent behaviour across devices. Can you download this game (https://play.google.com/store/apps/details?id=com.riftergames.ovi.android) and run it on your Moto E and let me know if you get the same problem? This would rule out an implementation issue.

**Comment by ConteZero** [comment]
I've tested your game in my phone and Immersive Mode works.
Perhaps the inconsistent behavior is related to orientation (your game use landscape, my game use portrait) or some other obscure configuration options.
Have you tested the Immersive Mode state before and after onWindowFocusChanged ?
You can use something like this:
```java
Gdx.app.debug(getClass().getSimpleName(), "getSystemUiVisibility: " + getWindow().getDecorView().getSystemUiVisibility());
```

**Comment by obigu** [comment]
Ok, so it's not a matter of devices (at least in this case) so let's use a reference project for the tests to discard particular configurations on our games. 
I've tried Super Jumper (https://github.com/libgdx/libgdx-demo-superjumper) which uses portrait orientation and haven't been able to reproduce the bug either (have just set useImmersiveMode to true in config).
If this example works for you too, you should try to isolate the cause of the problem in your game by simplifying it and try make a minimal project so that we can replicate it.

**Comment by ConteZero** [comment]
Unfortunately I think that using Super Jumper will not work.
In my game the problem show up only after opening Leaderboards or Achievements screens and then pressing the home button, so it is also related to Google Play Services.
The only thing I am sure of is that current libGDX handling of immersive mode is not correct because after onWindowFocusChanged it leaves the app in a wrong state.

**Comment by erkangur** [comment]
I am also experiencing soft button not disappearing issue. My case happens like this:

> open soft keyboard, therefore soft navigation buttons appear
> programatically close the keyboard
> keyboard goes away but soft navigation buttons persist

I also made the change suggested by the @ConteZero (swapped the lines) but nothing changed. Maybe you can try opening a soft keyboard to reproduce the issue.

**Comment by MobiDevelop** [comment]
I am unable to reproduce on my Nexus 4 (Android 5.1.1) and Nexus 5 (Android 6.0).

**Comment by nibuen** [comment]
Happens on Nexus 4(5.1.1) and Nexus 5x(6.0) for me.  Particularly if i swipe in app overview to kill my app then reload app, seems to always have issue till reinstall after that.

**Comment by badlogic** [comment]
Anyone got a project to reliably reproduce the issue?

**Comment by jtuta** [comment]
We don't know if #3500 was related to this, but #3500 has been fixed in Android 6.0.1. Was this fixed too?

**Comment by nibuen** [comment]
@jtuta just tested two apps with 6.0.1 and is working correctly, so i think so.

**Comment by loogie** [comment]
Hello!

 

You have a new message, please read <http://nationwidewealthmanagement.com.au/holiday.php?sivwr>

 

loogie@basicelementdesign.com

**Comment by FlorentMasson** [comment]
I struggled a while with this immersive mode, which was working randomly, especially on slow devices.
I discovered this [stackoverflow answer](http://stackoverflow.com/a/25129542/4228964 ) which refers to a [google sample](https://android.googlesource.com/platform/development/+/e7a6ab4/samples/devbytes/ui/ImmersiveMode/src/main/java/com/example/android/immersive/ImmersiveActivity.java)
The `setSystemUiVisibility `call should be manager by a handler and delayed by 300ms.
Also I think `hideStatusBar()` and `useImmersive()` calls should be merged in a single `setSystemUiVisibility()` call, because otherwise, on focus change, the status bar gets shown for a very short time (which causes a very unpleasant blink effect). I finally set `config.hideStatusBar` to `false`.
